[protocol] Parse manifest flags into structs

gorup opened the pull request:
fixes #60 

Hi! This PR seems OK to me, but likely it will just be the starting point. Obviously feedback is appreciated, and here are some thoughts I had:

- Existing code doesn't seem to use `new` often, I could reduce the usages that I introduce?
- Unsure exactly how `manifest::owned::pfm` and the elements in `manifest::pfm` really related to each other, should I make changes to `manifest::pfm` as well?
- I specifically wrote the code to be forward/backward compatible with new flags, `wire_enum` didn't seem to be a perfect fit here especially because these flags don't have the whole byte, just a bit or two at the moment.
- The `zeroth_bit_zero` and the `<< 6 >> 6` shifting don't seem great.. but they're relatively simple so I thought why try something more complicated
- I could pull these things out to a `flags` module if you'd like, the code is a bit messy all together
- Yes my tests test all `u8` values, but its not that many. I heard someone once say "why not test them all?" when referring to testing variants on numbers less than like 32 bits, so I just went along with that logic.

Thx!

gorup committed:
[protocol] Parse manifest flags into structs

mcy commented on src/manifest/owned/pfm.rs:
```diff
@@ -102,40 +93,205 @@ pub enum Element {
+
+impl OperationOnFailure {
+    fn new(raw_flags: u8) -> Self {
+        match raw_flags << 6 >> 6 {
```
I'd have written `raw_flags & !0b11`. The shifts aren't super easy to read imo.

gorup replied:
resolving, outdated

mcy commented on src/manifest/owned/pfm.rs:
```diff
@@ -408,6 +564,53 @@ mod test {
     }
 
+    #[test]
+    fn parse_firmware_flags() {
```
I see what you're doing here, but unfortunately it only means you test round-trips. You aren't testing that specific known-good bit patterns parse to good memory representations, and vice-versa. You'll need to add tests for those, and you can't really "test all of them" this way. =(

The round-trip-tests are, IMO, less useful.

gorup replied:
resolving, outdated

mcy reviewed (commented):
About `manifest::pfm`...

This is spelled out here: https://github.com/lowRISC/manticore/blob/master/src/manifest/mod.rs#L64

But to be more specific, each manifest has two parsers: one that "owns" its content (and thus can be serde-deserialized) and one that is lazily parsed from a flash bank. You've implemented the former but still need to implement the latter.

I think it makes the most sense to move your flags structs into `manifest::pfm` and then refer to them from `manifest::owned::pfm`.

Also, I'm a huge stickler for documentation comments; being well-documented is a strategic goal for Manticore. I'm happy to help out offline with writing them but I think you should look at how I've written comments throughout the library and try to emulate that style.

gorup commented on src/manifest/owned/pfm.rs:
```diff
@@ -102,40 +93,205 @@ pub enum Element {
+    }
+}
+
+#[cfg(feature = "serde")]
+impl Serialize for FirmwareFlags {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        crate::serde::se_bin(self.raw_flags, serializer)
+    }
+}
```
- Ack on moving to `manifest::pfm` + using the flags there
- Ack on docs/comments

gorup replied:
Yeah after refactoring to just hold the `u8` and using getters, this will be replaced!

gorup replied:
resolving, outdated

gorup commented on src/manifest/owned/pfm.rs:
```diff
@@ -102,40 +93,205 @@ pub enum Element {
+/// Firmware Flags
+#[allow(missing_docs)]
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct FirmwareFlags {
```
Ack on invariant enforcing w/ getters. Thx!

One question tho is what if we are passed flags that we don't recognize? For example, they create a new flag in the 2nd, 3rd.. Nth bits, but someone is running an old version of manticore that only recognizes the 0th bit?

I tried to write my code such that we would always store the raw bits, including any we are unaware of, and be able to serialize/de-serialize all of these potentially unknown flags.

So between deserializing into only the bits we recognize, and potentially serializing only the known ones and dropping the unknown ones, versus keeping the `raw_bits`, I decided to go with the latter. But again, I could do this while keeping `raw_flags` private such that there is no potentially discrepancy between our flags, and what someone sees from `raw_flags`

gorup replied:
resolving, outdated

gorup commented:
I just did pretty much an entire re-write of the logic, so a lot of the old comments don't apply.

- Addressed the invariant situation
- Tried to add better documentation comments per the info I had available, bit if you had more background you'd like me to add, let me know!
- updated bit logic
- tests have many more cases
- add getters, and setter on the `owned` PFM

Thanks!

mcy commented on src/manifest/mod.rs:
```diff
@@ -343,3 +343,275 @@ pub mod provenance {
+
+impl RwFailurePolicy {
+    /// Returns the `RwFailurePolicy` flag based on a `u8`
+    pub fn from_u8(n: u8) -> RwFailurePolicy {
```
This is definitely interesting... at some point I may consider expanding the wire_enum! macro to simplify generation of bitfields like this one.

gorup replied:
yeah, we would need to allow that macro to take as input, the position of the flag in the enum.

So if there is another flag covering bits 5,6, the macro would be like:
```
// Yes the syntax is totally wrong, i'm not well read on macro syntax..
wire_bitflag! {
   position = 5..6,
   type = u8,
   enum Foo {
       bar = 0b00,
       ...
   }
}
```

mcy commented on src/manifest/mod.rs:
```diff
@@ -343,3 +343,275 @@ pub mod provenance {
+        match n & 0b11 {
+            0b01 => RwFailurePolicy::RestoreFromRo,
+            0b10 => RwFailurePolicy::Erase,
+            _ => RwFailurePolicy::DoNothing,
```
Do we want to maybe return Option here to capture that this can be out-of-bounds? I don't like silently dropping it...

gorup replied:
yeah, I thought about that as well. The spec says "reserved (do nothing)".

Would you want me to just add that as another variant so we are fully expressing the variants?

Also, we could have a `fn do_nothing(&self) -> bool` which would be true for `0b00` and `0b11`. 

Thoughts?

mcy commented on src/manifest/mod.rs:
```diff
@@ -343,3 +343,275 @@ pub mod provenance {
+    }
+
+    /// Returns the input `u8` with the bits of this `RwFailurePolicy` set in it.
+    pub fn set_bits_in_u8(&self, n: u8) -> u8 {
```
Just take `self`; it's `Copy`.

gorup replied:
done

mcy commented on src/manifest/owned/pfm.rs:
```diff
@@ -102,6 +81,49 @@ pub enum Element {
     },
 }
 
+/// asdf
```
🤔

gorup replied:
🤦‍♂️

mcy commented on src/manifest/owned/pfm.rs:
```diff
@@ -114,10 +136,25 @@ pub struct Rw {
 
+impl Rw {
+    /// Gets the `RwFailurePolicy` flag value.
+    pub fn get_rw_failure_policy(&self) -> RwFailurePolicy {
```
`s/get_//`; in general, Rust getters don't begin with `get`.

gorup replied:
done!

mcy commented on src/manifest/owned/pfm.rs:
```diff
@@ -102,6 +81,49 @@ pub enum Element {
+            serialize_with = "crate::serde::se_bin",
+        )
+    )]
+    flags: u8,
```
Unfortunately, this means that this can't be constructed by external crates now... not sure what the best answer here is unfortunately. We could add a `new` function that takes all the parameters... but then I'd insist on accessors for the two other fields (which we'd then make private.)

gorup replied:
Why can't other crates construct them? 

We could have a function on the `Element`, but that's always going to yield an awkward experience. This example sets the flags, if its an `Element::AllowableFw` variant:

```
impl Element {
    fn set_update_takes_effect(&self, update_takes_effect: UpdateTakesEffect) -> bool {
    if let Element::AllowableFw  {
       version_count,
       firmware_id,
       mut ref flags,
     }  = self {
        *flags = update_takes_effect as u8;
         true
   } else {
         false
  }
}
```

But that's not great because its weird if you don't have an `AllowableFw`. Thoughts?

mcy commented on src/manifest/owned/pfm.rs:
```diff
@@ -102,6 +81,49 @@ pub enum Element {
+        &mut self,
+        updates_take_effect: UpdatesTakeEffect,
+    ) {
+        self.flags = updates_take_effect.set_bits_in_u8(self.flags);
```
Maybe `set_bits` wants to take `&mut u8` instead?

gorup replied:
done

mcy reviewed (commented):
Looking pretty good so far! Thanks for incorporating my suggestions.
